### ISSUE TITLE
Factor out storage key generation logic, to eliminate code duplication

### ISSUE THREAD

**Issue by ches** [feast-dev/feast#402]
This is a small-ish refactoring chore. See https://github.com/gojek/feast/pull/360#discussion_r362778747

Currently there is some code duplication between ingestion and serving for the generation and parsing of Redis keys. We multiplied this in the course of implementing Cassandra storage (#360), as it is already a large change set we didn't want to introduce additional Redis refactoring to it and thus decided to log it as a follow-up.

This could be done for Redis independent of and before integrating the Cassandra implementation though, and Cass implementation can be updated to follow the pattern if so.

_Update_: The below storage modularization was realized through #529 and subsequent PRs that implemented its interfaces. This issue remains for tech debt of code duplication that still remains, enumerated in https://github.com/gojek/feast/issues/402#issuecomment-623264345

~~In the interest of storage modularization to minimize dependency headaches in the future (something I believe we'll hash out in further RFC issues), I propose something like this:~~

~~1. A `storage-api` module, which defines an interface something like `KeyUtil`~~
~~1. A `storage-redis` implementation module which implements `KeyUtil` for Redis~~

~~I'm open to better naming suggestions, structure is more my focus here.~~

~~The `KeyUtil`s could ideally be `static` so that unit testing them is trivially simple and the unit tests can serve as a good spec/documentation of the key format.~~

**Comment by woop** [comment]
Makes a lot of sense. Tests that serve as the documentation would make this valuable.

**Comment by woop** [comment]
We should also be extending this to `feature reference` serialization and deserialization.

**Comment by ches** [comment]
> @ches would it make sense to broaden the scope here? Your proposal is currently focused on storage, which is what this issue is about, but there are various points of duplication throughout the code base that will only get worse.

Yes I definitely think it would make sense and that it's part of a larger modularization effort/need, I just wasn't prepared to write up a more comprehensively encompassing issue at the time I filed this one 😇  Wanted to put concrete pebbles into backlog consideration as I came across them, they may bubble up to the umbrella discussion and be low-hanging fruit/good-first-issue fodder in the meantime.

IMO this could be useful interim refactoring even if we refine what the end goal looks like as time goes on, it can be a progressive step without being counterproductive or a waste of time.

In short, I think this is one task within an epic.

**Comment by zhilingc** [comment]
@ches @woop 
I'm thinking of giving this a go sometime this week, hopefully with your blessing, haha. After looking around, I'm partial to [flink](https://github.com/apache/flink/tree/master/flink-connectors)'s implementation for multi-connector support, and propose something like this:

```
.
├── core
├── ingestion
├── serving
└── storage
    ├── api // interfaces, not abstract classes
    │   ├── BatchRetriever.java
    │   ├── FeatureStorage.java
    │   └── OnlineRetriever.java
    ├── common // Utils common across all stores that implementations may or may not use
    │   └── retry // existing retry code
    │       ├── BackOffExecutor.java
    │       └── Retriable.java
    └── connectors // implementations
        ├── bigquery
        ├── cassandra
        └── redis
            ├── RedisFeatureStorage.java
            ├── RedisKey.java
            └── RedisOnlineRetriever.java
```

I'm thinking of keeping the KeyUtil separate from the API since it's specific to KV stores and not really a general need, so I don't think Feast should be opinionated about their implementation. Connectors are free to share a single KeyUtil across their Storage and Retriever implementations.

The actual methods defined in the interfaces require further thought, but just throwing this out for comments first.

**Comment by ches** [comment]
LGTM!

> I'm thinking of keeping the KeyUtil separate from the API since it's specific to KV stores and not really a general need, so I don't think Feast should be opinionated about their implementation.

True, though I think it's fine to have interfaces in `api` that not every implementation needs, but on the other hand should be conservative about promoting things there until their abstractions have some exercise.

Hypothetical KeyUtil might not end up being a good case though anyway, among other things I had thought they'd ideally be static and there's really no way to have static interfaces in Java, so maybe that's a clue it's against the grain… (not to mention "util" usually makes me feel a little ashamed of myself 😅).

> Connectors are free to share a single KeyUtil across their Storage and Retriever implementations.

Definitely intended this to be true though, as the original spirit of this issue.

**Comment by woop** [comment]
@ches @zhilingc Just want to confirm the state of this issue. It's slated for 0.5.0. Either we should decouple it from 0.5.0, close it, or add more work to our backlog.

Can it be closed?

**Comment by ches** [comment]
So we sort of let the scope of this issue morph over time to representing the broader goals of [storage modularization](https://github.com/gojek/feast/projects/6), but I would say the original intent of this issue still isn't resolved until:

- These are deduped:
  - https://github.com/gojek/feast/blob/3d9bafd12515b872c479cfb0f7369e11eb0663d9/serving/src/main/java/feast/serving/util/RefUtil.java#L39
  - https://github.com/gojek/feast/blob/3d9bafd12515b872c479cfb0f7369e11eb0663d9/storage/connectors/redis/src/main/java/feast/storage/connectors/redis/retriever/RedisOnlineRetriever.java#L214
  - https://github.com/gojek/feast/blob/3d9bafd12515b872c479cfb0f7369e11eb0663d9/storage/connectors/rediscluster/src/main/java/feast/storage/connectors/rediscluster/retriever/RedisClusterOnlineRetriever.java#L219
- And these:
  - https://github.com/gojek/feast/blob/3d9bafd12515b872c479cfb0f7369e11eb0663d9/storage/connectors/redis/src/main/java/feast/storage/connectors/redis/retriever/RedisOnlineRetriever.java#L120
  - https://github.com/gojek/feast/blob/3d9bafd12515b872c479cfb0f7369e11eb0663d9/storage/connectors/rediscluster/src/main/java/feast/storage/connectors/rediscluster/retriever/RedisClusterOnlineRetriever.java#L127
- And these:
  - https://github.com/gojek/feast/blob/3d9bafd12515b872c479cfb0f7369e11eb0663d9/storage/connectors/redis/src/main/java/feast/storage/connectors/redis/retriever/RedisOnlineRetriever.java#L99
  - https://github.com/gojek/feast/blob/3d9bafd12515b872c479cfb0f7369e11eb0663d9/storage/connectors/rediscluster/src/main/java/feast/storage/connectors/rediscluster/retriever/RedisClusterOnlineRetriever.java#L106

The original title and most of the description of the issue still apply for cleaning that up I'd say. So my vote would be keep this issue, but remove it from v0.5.0 milestone if that tech debt is low priority.

**Comment by ches** [comment]
I can update the description if that's how we go.

**Comment by woop** [comment]
> So my vote would be keep this issue, but remove it from v0.5.0 milestone if that tech debt is low priority.

Let's do this please :)
